server/server.go

```diff
@@ -257,42 +218,52 @@ func NewServer(cfg *config.Config, driver IDriver) (*Server, error) {
 		if s.cfg.EnableTCP4Only {
 			tcpProto = "tcp4"
 		}
-		if s.listener, err = net.Listen(tcpProto, addr); err == nil {
-			logutil.BgLogger().Info("server is running MySQL protocol", zap.String("addr", addr))
-			if cfg.Socket != "" {
-				if s.socket, err = net.Listen("unix", s.cfg.Socket); err == nil {
-					logutil.BgLogger().Info("server redirecting", zap.String("from", s.cfg.Socket), zap.String("to", addr))
-					go s.forwardUnixSocketToTCP()
-				}
-			}
-			if runInGoTest && s.cfg.Port == 0 {
-				s.cfg.Port = uint(s.listener.Addr().(*net.TCPAddr).Port)
-			}
+		if s.listener, err = net.Listen(tcpProto, addr); err != nil {
+			return nil, errors.Trace(err)
 		}
-	} else if cfg.Socket != "" {
-		if s.listener, err = net.Listen("unix", cfg.Socket); err == nil {
-			logutil.BgLogger().Info("server is running MySQL protocol", zap.String("socket", cfg.Socket))
+		logutil.BgLogger().Info("server is running MySQL protocol", zap.String("addr", addr))
+		if runInGoTest && s.cfg.Port == 0 {
+			s.cfg.Port = uint(s.listener.Addr().(*net.TCPAddr).Port)
 		}
-	} else {
+	}
+
+	if s.cfg.Socket != "" {
+		if s.socket, err = net.Listen("unix", s.cfg.Socket); err != nil {
+			return nil, errors.Trace(err)
+		}
+		logutil.BgLogger().Info("server is running MySQL protocol", zap.String("socket", s.cfg.Socket))
+	}
+
+	if s.socket == nil && s.listener == nil {
 		err = errors.New("Server not configured to listen on either -socket or -host and -port")
+		return nil, errors.Trace(err)
 	}
 
-	if cfg.ProxyProtocol.Networks != "" {
-		pplistener, errProxy := proxyprotocol.NewListener(s.listener, cfg.ProxyProtocol.Networks,
-			int(cfg.ProxyProtocol.HeaderTimeout))
-		if errProxy != nil {
+	if s.cfg.ProxyProtocol.Networks != "" {
```

Comment:
Maybe this needs to be executed twice, in `startNetworkListener`.

You see, you did not block user from passing `socket` and `host:port` simultaneously. So it is possible to start two listeners, while only one get the proxy applied.

Reply:
I tried to keep the current behaviour, where it only would add a proxy on the s.listener (which is the host:port network interface if specified otherwise s.listener would be set to the unix domain socket).

There may be a use case for supporting proxy protocol for unix socket independently of the network interface, but I think we can wait until someone requests it. I would imagine that it could remove some overhead if there is a proxy network process on the same server, but in this patch I prefer to keep the behaviour for this feature.

Reply:
OK, I am fine with it, then.

Reply:
BTW, you may want to talk about the functionality of starting two listeners, it is a new function if you don't forbid it.

Reply:
https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

Reply:
> BTW, you may want to talk about the functionality of starting two listeners, it is a new function if you don't forbid it.

The previous code also allowed two listeners, one on network interface(s) and one on Unix Domain Socket, so no change in functionality I think.

Reply:
> https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

Not sure if you just wanted to help or point something out? It looks like the proxy protocol do support AF_UNIX / Unix Domain Sockets, but I don't think we should spend time on adding it to this issue, probably wait until there are customer requests.